File: scripts/generate_parity.py
import os
import re

def get_files(directory):
    file_list = []
    if not os.path.exists(directory): return file_list
    for root, _, files in os.walk(directory):
        for file in files:
            file_list.append(os.path.join(root, file))
    return file_list

def extract_basename(filename):
    base = os.path.basename(filename)
    base = re.sub(r'\.(component|service|directive|pipe)\.ts$', '', base)
    base = re.sub(r'\.(dart|ts)$', '', base)
    base = re.sub(r'_(widget|screen|bloc|provider|model|types|cubit|event|service)$', '', base)
    base = re.sub(r'(-|_)', ' ', base)
    return base.lower().strip()

script_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(script_dir)

angular_root = os.path.join(project_root, 'src')
flutter_root = os.path.join(project_root, 'pocketgull_flutter', 'lib')

angular_dirs = ['components', 'services', 'directives', 'pipes']
flutter_dirs = ['widgets', 'screens', 'services', 'providers', 'models', 'theme']

angular_files = []
for d in angular_dirs:
    angular_files.extend(get_files(os.path.join(angular_root, d)))
angular_files = [f for f in angular_files if not f.endswith('.spec.ts')]

flutter_files = []
for d in flutter_dirs:
    flutter_files.extend(get_files(os.path.join(flutter_root, d)))

angular_map = {}
for f in angular_files:
    base = extract_basename(f)
    if base not in angular_map: angular_map[base] = []
    angular_map[base].append(os.path.relpath(f, project_root).replace('\\', '/'))

flutter_map = {}
for f in flutter_files:
    base = extract_basename(f)
    if base not in flutter_map: flutter_map[base] = []
    flutter_map[base].append(os.path.relpath(f, project_root).replace('\\', '/'))

all_keys = sorted(list(set(angular_map.keys()) | set(flutter_map.keys())))

markdown = '# Feature Parity Matrix\n\n'
markdown += 'This document maps the components and services from the live Angular site to the new Flutter migration to track feature parity.\n\n'
markdown += '| Feature / Base Name | Angular (Live) | Flutter (Migration) | Status |\n'
markdown += '| :--- | :--- | :--- | :--- |\n'

match_count = 0
missing_flutter = 0
flutter_only = 0

for key in all_keys:
    ang_files = angular_map.get(key, [])
    flt_files = flutter_map.get(key, [])
    
    status = '✅ Parity'
    if len(ang_files) > 0 and len(flt_files) == 0:
        status = '❌ Missing in Flutter'
        missing_flutter += 1
    elif len(ang_files) == 0 and len(flt_files) > 0:
        status = '⚠️ Flutter Only'
        flutter_only += 1
    else:
        match_count += 1
        
    ang_str = '<br>'.join([f'`{f}`' for f in ang_files])
    flt_str = '<br>'.join([f'`{f}`' for f in flt_files])
    
    markdown += f"| **{key}** | {ang_str if ang_str else '-'} | {flt_str if flt_str else '-'} | {status} |\n"

markdown += '\n## Summary\n'
markdown += f"- **Matched Features**: {match_count}\n"
markdown += f"- **Missing in Flutter (Needs Migration)**: {missing_flutter}\n"
markdown += f"- **Flutter Only (New Features/Architecture)**: {flutter_only}\n"

output_path = os.path.join(project_root, 'parity_matrix.md')
with open(output_path, 'w', encoding='utf-8') as f:
    f.write(markdown)

print(f"Parity matrix generated at {output_path}")

File: scripts/dart/generate_parity.dart
import 'dart:io';

String findProjectRoot() {
  var dir = File(Platform.script.toFilePath()).parent;
  while (dir.path != dir.parent.path) {
    final pkgJson = File('${dir.path}${Platform.pathSeparator}package.json');
    if (pkgJson.existsSync()) {
      return dir.path;
    }
    dir = dir.parent;
  }
  return Directory.current.path;
}

List<File> getFiles(Directory dir) {
  if (!dir.existsSync()) return [];
  return dir.listSync(recursive: true).whereType<File>().toList();
}

String extractBasename(String filePath) {
  var base = filePath.split(Platform.pathSeparator).last.split('/').last;
  base = base.replaceAll(RegExp(r'\.(component|service|directive|pipe)\.ts$'), '');
  base = base.replaceAll(RegExp(r'\.(dart|ts)$'), '');
  base = base.replaceAll(RegExp(r'_(widget|screen|bloc|provider|model|types|cubit|event|service)$'), '');
  base = base.replaceAll(RegExp(r'[-_]'), ' ');
  return base.toLowerCase().trim();
}

void main() {
  print('\n⚖️  Pocket-Gull Angular <-> Flutter Feature Parity Auditor (Dart)\n');

  final root = findProjectRoot();
  final sep = Platform.pathSeparator;

  final angularRoot = Directory('$root${sep}src');
  final flutterRoot = Directory('$root${sep}pocketgull_flutter${sep}lib');

  final angularDirs = ['components', 'services', 'directives', 'pipes'];
  final flutterDirs = ['widgets', 'screens', 'services', 'providers', 'models', 'theme'];

  final angularFiles = <File>[];
  for (final d in angularDirs) {
    angularFiles.addAll(getFiles(Directory('${angularRoot.path}$sep$d')));
  }
  final cleanAngularFiles = angularFiles.where((f) => !f.path.endsWith('.spec.ts')).toList();

  final flutterFiles = <File>[];
  for (final d in flutterDirs) {
    flutterFiles.addAll(getFiles(Directory('${flutterRoot.path}$sep$d')));
  }

  final angularMap = <String, List<String>>{};
  for (final f in cleanAngularFiles) {
    final base = extractBasename(f.path);
    final relPath = f.path.replaceFirst('$root$sep', '').replaceAll('\\', '/');
    angularMap.putIfAbsent(base, () => []).add(relPath);
  }

  final flutterMap = <String, List<String>>{};
  for (final f in flutterFiles) {
    final base = extractBasename(f.path);
    final relPath = f.path.replaceFirst('$root$sep', '').replaceAll('\\', '/');
    flutterMap.putIfAbsent(base, () => []).add(relPath);
  }

  final allKeys = {...angularMap.keys, ...flutterMap.keys}.toList()..sort();

  final buffer = StringBuffer();
  buffer.writeln('# Feature Parity Matrix\n');
  buffer.writeln('This document maps the components and services from the live Angular application to the Flutter companion suite to track architectural parity.\n');
  buffer.writeln('| Feature / Base Name | Angular (Live) | Flutter (Companion) | Status |');
  buffer.writeln('| :--- | :--- | :--- | :--- |');

  var matchCount = 0;
  var missingFlutter = 0;
  var flutterOnly = 0;

  for (final key in allKeys) {
    final angFiles = angularMap[key] ?? [];
    final fltFiles = flutterMap[key] ?? [];

    String status;
    if (angFiles.isNotEmpty && fltFiles.isEmpty) {
      status = '❌ Missing in Flutter';
      missingFlutter++;
    } else if (angFiles.isEmpty && fltFiles.isNotEmpty) {
      status = '⚠️ Flutter Only';
      flutterOnly++;
    } else {
      status = '✅ Parity';
      matchCount++;
    }

    final angStr = angFiles.isNotEmpty ? angFiles.map((f) => '`$f`').join('<br>') : '-';
    final fltStr = fltFiles.isNotEmpty ? fltFiles.map((f) => '`$f`').join('<br>') : '-';

    buffer.writeln('| **$key** | $angStr | $fltStr | $status |');
  }

  buffer.writeln('\n## Summary');
  buffer.writeln('- **Matched Features**: $matchCount');
  buffer.writeln('- **Missing in Flutter (Needs Migration)**: $missingFlutter');
  buffer.writeln('- **Flutter Only (New Architecture/Components)**: $flutterOnly');

  final outputPath = '$root${sep}parity_matrix.md';
  final docsPath = '$root${sep}docs${sep}research${sep}parity_matrix.md';
  File(outputPath).writeAsStringSync(buffer.toString());
  if (File(docsPath).existsSync()) {
    File(docsPath).writeAsStringSync(buffer.toString());
  }

  print('  [OK] Matched Features: $matchCount');
  print('  [OK] Missing in Flutter: $missingFlutter');
  print('  [OK] Flutter Only: $flutterOnly');
  print('  [OK] Parity matrix generated at: $outputPath');
  print('  [OK] Parity matrix mirrored at: $docsPath');
  print('\n✅ Parity audit complete!\n');
}
